identity-account-storage/src/types/encryption_algorithm.rs

```diff
@@ -0,0 +1,11 @@
+// Copyright 2020-2022 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use serde::Deserialize;
+use serde::Serialize;
+
+#[non_exhaustive]
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub enum EncryptionAlgorithm {
+  Aes256Gcm,
+}
```

Comment:
Specifying `Aes256Gcm` alone might be insufficient. There are basically three configurable parts to encryption:

- Key exchange algorithm (e.g. RSA, ECDH), which is implicit by the type of key used.
- Encryption algorithm (e.g. AES256-GCM, AES256-CBC + HMAC-SHA512).
- Key wrapping algorithm (e.g. A256KW).

Unfortunately it's not so clear-cut. We have to deal with algorithms like ECDH-1PU which has additional steps (performs two Diffie-Hellman key exchanges using the ephemeral and static keys) and inputs.

We might want to follow the approach taken in `libjose`, where the entire algorithm is specified as an enum, e.g. `ECDH_1PU_A256KW`, except with a rich enum for additional parameters required.

https://github.com/iotaledger/identity.rs/blob/4287c9834de8dc42fd4ff0c0e8b3a2bc822d0ec4/libjose/src/jwe/algorithm.rs#L95

Note that this is separate from the encryption algorithm, which could be specified separately?
https://github.com/iotaledger/identity.rs/blob/4287c9834de8dc42fd4ff0c0e8b3a2bc822d0ec4/libjose/src/jwe/encryption.rs#L39

Basically, all this is saying is we need more flexibility/parameters/enums than just `EncryptionAlgorithm` to determine encryption behaviour.

Reply:
I made a similar addition to `identity-comm` for specifying the algorithm to determine the content encryption key (CEK) (called `CEKAlgorithm`) and that in addition to the `EncryptionAlgorithm` seems to work well so far, but I can't say yet whether that is the universally the best design. It seems like a good start though, here it is for reference:

https://github.com/iotaledger/identity.rs/blob/dbdda47160e8725424a31dfd0ff4a274f323b3f6/identity-comm/src/envelope/encrypted.rs#L131-L167